### ISSUE TITLE
[New Rule] Exceptions MUST NOT be handled in the same function where they are thrown

### ISSUE THREAD

**Issue by lenaorobei** [magento/magento-coding-standard#29]
### Rule
Exceptions MUST NOT be handled in the same function where they are thrown.

### Reason
Source: [Magento Technical Guidelines](https://devdocs.magento.com/guides/v2.3/coding-standards/technical-guidelines.html).

### Implementation
Find `try-catch` block and `throw` keyword in the same method.
Exception type must be the same.

### Addtional Information 
**Type:**  warning
**Severity:**  8

**Comment by paliarush** [comment]
I believe we should have one more restriction, exception type must be the same.

**Comment by larsroettig** [comment]
Test Case Examples:

```
<?php


class CustomFooException extends Exception
{

}


class Foo
{
    /**
     * @param array $data
     * @return array|mixed
     */
    private function vaildateFunction(array $data)
    {
        try {
            if (!isset($data['formData'])) {
                throw new \CustomFooException('Bar is not set.');
            }
            return $data['formData'];
        } catch (CustomFooException $exception) {
            // handle $exception code
        }

        return [];
    }

    /**
     * @param array $data
     */
    public function complexDivisionFunction(array $data)
    {
        try {

            $result = $this->vaildateFunction($data);

            $previous = 1;

            foreach ($result as $number) {
                if ($number === 0) {
                    throw new LogicException('Cant not deviated by null');
                }

                $previous /= $number;
            }

            return $previous;
        } catch (CustomFooException $exception) {
            // handle $exception code
        } catch (\LogicException $logicException) {
            // handle $exception code
        }
    }
}


```

**Comment by lenaorobei** [comment]
Implemented in magento/magento-coding-standard#43
